File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.js
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];

readline.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  const result = solution(input);
  console.log(result);
  process.exit();
});

const solution = (input) => {
  const N = Number(input[0]);
  const fruits = input[1].split(' ').map(Number);
  const types = {};

  if (N === 1) return 1;

  if (!types[fruits[0]]) types[fruits[0]] = 0;
  types[fruits[0]] += 1;
  if (!types[fruits[1]]) types[fruits[1]] = 0;
  types[fruits[1]] += 1;

  let result = Object.keys(types).length;
  let left = 0, right = 1;
  while (right < N) {
    if (Object.keys(types).length > 2) { 
      types[fruits[left]] -= 1; 
      if (types[fruits[left]] === 0) delete types[fruits[left]];
      left++; 
    } else {
      if (!types[fruits[right+1]]) types[fruits[right+1]] = 0;
      types[fruits[right+1]] += 1;
      right++;

      if (Object.keys(types).length <= 2 && right < N) {
        result = Math.max(result, right-left+1);
      }
    }
  }

  return result;
}

File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.py
import sys 
input = sys.stdin.readline

N = int(input())
fruits = list(map(int, input().split())) 

fruitDict = {}
fruitDict[fruits[0]] = 1

if N == 1:
  print(1)
else:
  if fruits[1] in fruitDict: fruitDict[fruits[1]] += 1
  else: fruitDict[fruits[1]] = 1

  maxResult = 1
  left = 0
  right = 1
  while left <= right and right < N:
    if len(fruitDict) <= 2:
      maxResult = max(maxResult, right-left+1)
      right += 1
      if (right < N):
        if fruits[right] in fruitDict: fruitDict[fruits[right]] += 1
        else: fruitDict[fruits[right]] = 1
    else:
      fruitDict[fruits[left]] -= 1
      if fruitDict[fruits[left]] <= 0: del fruitDict[fruits[left]]
      left += 1

  print(maxResult)